ocaml/xapi/xapi_pool_update.ml

```diff
@@ -524,38 +531,42 @@ let introduce ~__context ~vdi =
 let pool_apply ~__context ~self =
   let pool_update_name = Db.Pool_update.get_name_label ~__context ~self in
   debug "pool_update.pool_apply %s" pool_update_name ;
+  let module HostSet = Set.Make (struct
+    type t = API.ref_host
+
+    let compare = compare
+  end) in
   let unapplied_hosts =
     Db.Pool_update.get_hosts ~__context ~self
-    |> List.set_difference (Db.Host.get_all ~__context)
+    |> HostSet.of_list
+    |> HostSet.diff (Db.Host.get_all ~__context |> HostSet.of_list)
   in
-  if List.length unapplied_hosts = 0 then (
+  if HostSet.is_empty unapplied_hosts then (
     debug "pool_update.pool_apply, %s has already been applied on all hosts."
       pool_update_name ;
     raise
       (Api_errors.Server_error
          (Api_errors.update_already_applied_in_pool, [Ref.string_of self]))
   ) else
     let failed_hosts =
-      unapplied_hosts
-      |> List.fold_left
-           (fun acc host ->
-             try
-               ignore
-                 (Helpers.call_api_functions ~__context (fun rpc session_id ->
-                      Client.Pool_update.apply rpc session_id self host)) ;
-               acc
-             with e ->
-               debug "Caught exception while pool_apply %s: %s"
-                 (Ref.string_of host)
-                 (ExnHelper.string_of_exn e) ;
-               host :: acc)
-           []
+      HostSet.fold
```

Comment:
I'm a bit surprised about the `ignore` inside a `fold`. Could this be done with `HashSet.iter` instead?

Reply:
It appears to accumulate errors, so a straight iter wouldn't work, but I agree that `ignore` inside a fold is awkward.
Perhaps a clearer way to write this would be to List.map the outcome to a result, then filter out only the errors from that list. We can refactor this on master, would keep the backport identical to master for now.